### PR TITLE
i18n support

### PR DESCRIPTION
**Pull request by vishalnarkhede**
# Submit a pull request

## CLA

- [ ] I have signed the [Stream CLA](https://docs.google.com/forms/d/e/1FAIpQLScFKsKkAJI7mhCr7K9rEIOpqIDThrWxuvxnwUq2XkHyG154vQ/viewform) (required).
- [ ] The code changes follow best practices
- [ ] Code changes are tested (add some information if not applicable)

### Currently available on rc version - [0.7.0-i18n-rc.2](https://github.com/GetStream/stream-chat-react-native/releases/tag/v0.7.0-i18n-rc.2)

Streami18n - Wrapper around [i18next](https://www.i18next.com/) class for Stream related translations.

### API

- **constructor**(options)

  Contructor accepts following options:

  - **language** (String) default: 'en'

    Language code e.g., en, tr

  - **disableDateTimeTranslations** (boolean) default: false

    Disable translations for datetimes

  - **debug** (boolean) default: false

    Enable debug mode in internal i18n class

  - **logger** (function) default: () => {}

    Logger function to log warnings/errors from this class

  - **momentLocaleConfigForLanguage** (object) default: 'enConfig'

    [Config object](https://momentjs.com/docs/#/i18n/changing-locale/) for internal moment object, corresponding to language (param)

- **geti18Instance**

  Returns an instance of [i18next](https://www.i18next.com/) used internally.

- **getAvailableLanguages**

  Returns all the languages (code) registered with Streami18n

- **getTranslations**

  Returns all the translations registered with Streami18n

- **getTranslators**

  Returns an object containing t (i18next translator) and momentjs instance (configured with set language)

  ```js static
  const streami18n = new Streami18n({ language: 'nl' });
  const { t, moment } = await streami18n.getTranslators();
  ```

- **registerTranslation**

  _params_

  - language | string
  - translator | object
  - customMomentLocale | object (optional)

  ```js static
    streami18n.registerTranslation(
      'mr',
      {
        'Nothing yet...': 'काहीही नाही  ...',
        '{{ firstUser }} and {{ secondUser }} are typing...':
            '{{ firstUser }} आणि {{ secondUser }} टीपी करत आहेत ',
      },
      {
        months: [...],
        monthsShort: [...],
        calendar: {
            sameDay: '...'
        }
      }
    );
  ```

- **setLanguage**

  Set a different language

  ```js static
  streami18n.setLanguage('tr');
  ```

Instance of this class should be provided to Chat component to handle translations.
Stream provides following list of in-built translations:

1. English (en)
2. Dutch (nl)
3. Russian (ru)
4. Turkish (tr)
5. French (fr)
6. Italian (it)
7. Hindi (hi)

### Docs

- **Text translations**

  Simplest way to start using chat components in one of the in-built languages would be following:

  ```js static
  const i18n = new Streami18n({ language: 'nl' });
  <Chat client={chatClient} i18nInstance={i18n}>
    ...
  </Chat>;
  ```

  If you would like to override certain keys in in-built translation.
  UI will be automatically updated in this case.

  ```js static
  const i18n = new Streami18n({
    language: 'nl',
    translationsForLanguage: {
      'Nothing yet...': 'Nog Niet ...',
      '{{ firstUser }} and {{ secondUser }} are typing...':
        '{{ firstUser }} en {{ secondUser }} zijn aan het typen...',
    },
  });
  ```

  If you would like to register additional languages, use registerTranslation. You can add as many languages as you want:

  ```js static
  i18n.registerTranslation('zh', {
    'Nothing yet...': 'Nog Niet ...',
    '{{ firstUser }} and {{ secondUser }} are typing...':
      '{{ firstUser }} en {{ secondUser }} zijn aan het typen...',
  });

  <Chat client={chatClient} i18nInstance={i18n}>
    ...
  </Chat>;
  ```

  You can use the same function to add whole new language as well.

  ```js static
  const i18n = new Streami18n();

  i18n.registerTranslation('mr', {
    'Nothing yet...': 'काहीही नाही  ...',
    '{{ firstUser }} and {{ secondUser }} are typing...':
      '{{ firstUser }} आणि {{ secondUser }} टीपी करत आहेत ',
  });

  // Make sure to call setLanguage to reflect new language in UI.
  i18n.setLanguage('it');

  <Chat client={chatClient} i18nInstance={i18n}>
    ...
  </Chat>
  ```

- **Datetime translations**

  Stream components uses [momentjs](http://momentjs.com/) internally to format datetime stamp. e.g., in ChannelPreview, MessageContent components.
  Momentjs has locale support as well -https://momentjs.com/docs/#/i18n/

  Momentjs provides locale config for plenty of languages, you can check the whole list of locale configs at following url
  https://github.com/moment/moment/tree/develop/locale

  You can either configure locale for moment globally or you can provide the locale config while registering
  language with Streami18n (either via constructor or registerTranslation())

  1. Via import

    Easiest way to register a locale with momentjs is via import in your app.
    ```js static
    import 'moment/locale/nl';
    import 'moment/locale/it';
    ```

  2. Via language registration

    You can provide custom locale config for moment, while registering a language with Streami18n

    e.g.,
    ```js static
    const i18n = new Streami18n({
    language: 'nl',
    momentLocaleConfigForLanguage: {
      months: [...],
      monthsShort: [...],
      calendar: {
        sameDay: ...'
      }
    }
    });
    ```
    Similarly, you can add locale config for moment while registering translation via `registerTranslation` function.
    e.g.,

    ```js static
      const i18n = new Streami18n();
      i18n.registerTranslation(
      'mr',
      {
        'Nothing yet...': 'काहीही नाही  ...',
        '{{ firstUser }} and {{ secondUser }} are typing...': '{{ firstUser }} आणि {{ secondUser }} टीपी करत आहेत ',
      },
      {
        months: [...],
        monthsShort: [...],
        calendar: {
          sameDay: ...'
        }
      }
      );
    ```

**TIP**: If you would like to stick with english language for datetimes in Stream compoments,
you can set `disableDateTimeTranslations` to true.

```js static
const i18n = new Streami18n({
  language: 'nl',
  disableDateTimeTranslations: false
});

```

The default `en` locale config from moment is as follow:
```json static
{
  "months": [
    "January",
    "February",
    "March",
    "April",
    "May",
    "June",
    "July",
    "August",
    "September",
    "October",
    "November",
    "December"
  ],
  "monthsShort": [
    "Jan",
    "Feb",
    "Mar",
    "Apr",
    "May",
    "Jun",
    "Jul",
    "Aug",
    "Sep",
    "Oct",
    "Nov",
    "Dec"
  ],
  "week": {
    "dow": 0,
    "doy": 6
  },
  "weekdays": [
    "Sunday",
    "Monday",
    "Tuesday",
    "Wednesday",
    "Thursday",
    "Friday",
    "Saturday"
  ],
  "weekdaysMin": ["Su", "Mo", "Tu", "We", "Th", "Fr", "Sa"],
  "weekdaysShort": ["Sun", "Mon", "Tue", "Wed", "Thu", "Fri", "Sat"],
  "calendar": {
    "sameDay": "[Today at] LT",
    "nextDay": "[Tomorrow at] LT",
    "nextWeek": "dddd [at] LT",
    "lastDay": "[Yesterday at] LT",
    "lastWeek": "[Last] dddd [at] LT",
    "sameElse": "L"
  },
  "longDateFormat": {
    "LTS": "h:mm:ss A",
    "LT": "h:mm A",
    "L": "MM/DD/YYYY",
    "LL": "MMMM D, YYYY",
    "LLL": "MMMM D, YYYY h:mm A",
    "LLLL": "dddd, MMMM D, YYYY h:mm A"
  },
  "invalidDate": "Invalid date",
  "ordinal": "%d.",
  "dayOfMonthOrdinalParse": /\\d{1,2}(th|st|nd|rd)/,
  "relativeTime": {
    "future": "in %s",
    "past": "%s ago",
    "s": "a few seconds",
    "ss": "%d seconds",
    "m": "a minute",
    "mm": "%d minutes",
    "h": "an hour",
    "hh": "%d hours",
    "d": "a day",
    "dd": "%d days",
    "M": "a month",
    "MM": "%d months",
    "y": "a year",
    "yy": "%d years"
  },
  "meridiemParse": /[ap]\\.?m?\\.?/i,
  "abbr": "en"
}
```

### CODE REVIEW

File: bin/validate-translations.js [hunk 1]
```diff
@@ -0,0 +1,35 @@
+/* eslint-disable no-undef */
```
Comment (vishalnarkhede) [inline bin/validate-translations.js]: Ignore this file for now!!

File: .babelrc [hunk 2]
```diff
@@ -1,6 +1,22 @@
 {
   "presets": ["@babel/env", "@babel/react"],
   "plugins": [
+    [
+      "i18next-extract",
```
Comment (vishalnarkhede) [inline .babelrc]: This plugin scans through codebase and generates the translation files for all the supported languages!!

Comment (vishalnarkhede) [reply .babelrc]: So lets say if I add a new component somewhere `<Text>t('Vishal is awesome')</Text>`.

Plugin will append this key to every language file.

File: bin/validate-translations.js (hunk 1 above)
Comment (vishalnarkhede) [inline bin/validate-translations.js]: Small script to make sure all the keys in translation files have translation!

**Comment by jaapbakker88** [comment]
Tested this today. Looks really good 👍 

Couple things might be nice:
- Use default language and show warning as yellow box(in development) if language doesnt exist
- It would be nice if we could find a way to localise the dates. Moment can do this but it might be nice not to include all of moments languages by default.

File: README.md [hunk 3]
```diff
@@ -291,3 +291,147 @@ In current context, dependencies such as `react-native-document-picker` and (if
...
+
+  ```js static
+
+    // Registering locale config for Danish language
+    import 'moment/locale/da';
+
+    const i18n = new Streami18n({
+      language: 'da',
+      translationsForLanguage: {
+        ...
```
Comment (jaapbakker88) [inline README.md]: Does this automatically use moments date locales

File: README.md [hunk 4]
```diff
@@ -291,3 +291,147 @@ In current context, dependencies such as `react-native-document-picker` and (if
...
+  'mr',
+  {
+    'Nothing yet...': 'काहीही नाही  ...',
+    '{{ firstUser }} and {{ secondUser }} are typing...': '{{ firstUser }} आणि {{ secondUser }} टीपी करत आहेत ',
+  },
+  {
+    months: [...],
+    monthsShort: [...],
+    calendar: {
+      sameDay: '...'
```
Comment (jaapbakker88) [inline README.md]: Not sure if this is a clear name, does moment use the same name for this?

**Pull request merged**